### PR TITLE
feat(quantic): Create QuanticRefineModalContent component

### PR DESCRIPTION
**Pull request by mmitiche**
[SFINT-4471](https://coveord.atlassian.net/browse/SFINT-4471)
The `QuanticRefineModalContent` component duplicates and displays the facets present in the search interface in addition to the sort component. This component is primarily  made to be displayed inside a modal to assure the responsiveness  when the search interface is displayed in smaller screens.

#### Demo

https://user-images.githubusercontent.com/86681870/165296127-f7d82921-cff9-4345-a5fe-1e151431e2e0.mp4

### CODE REVIEW

File: packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.js
Comment (lbergeron) [inline packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.js:194]: You could use `reduce` here to copy a subset of the object properties.

```suggestion
    return properties.reduce((data, property) => {
      data[property] = element[property];
      return data;
    }, {});
```

File: packages/quantic/cypress/integration/refine-modal-content/refine-modal-content-expectations.ts
Comment (mmitiche) [reply packages/quantic/cypress/integration/refine-modal-content/refine-modal-content-expectations.ts]: Updated 👌

File: packages/quantic/cypress/integration/refine-modal-content/refine-modal-content-expectations.ts
Comment (mmitiche) [reply packages/quantic/cypress/integration/refine-modal-content/refine-modal-content-expectations.ts]: Yes I was able to do that using the [eql method](https://www.chaijs.com/api/bdd/#method_eql) 👌

File: packages/quantic/cypress/integration/refine-modal-content/refine-modal-content-selectors.ts
Comment (mmitiche) [reply packages/quantic/cypress/integration/refine-modal-content/refine-modal-content-selectors.ts]: Neat idea! I followed the logic you described to implement it, thank you!

File: packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.js
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.js:190]: updated 👌

File: packages/quantic/cypress/integration/refine-modal-content/refine-modal-content-selectors.ts
Comment (lbergeron) [inline packages/quantic/cypress/integration/refine-modal-content/refine-modal-content-selectors.ts:32]: You could get rid of the ternary on each selector by passing a Cypress selector to `getCommonFacetSelectors` and chaining the other calls from there.

For example:
```
const getCommonFacetSelectors = (rootSelector): FacetSelector => {
  return {
    get: () => rootSelector(),
    facet: () => rootSelector().find('c-quantic-facet'),
    ...
  };
}
```

and you would invoke `getCommonFacetSelectors` this way:
```
getCommonFacetContainer(() => cy.get(refineContentComponent));
getCommonFacetContainer(() => cy.get(refineContentContainer));
```

File: packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.js
Comment (lbergeron) [inline packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.js:120]: I think these functions don't have to be `static`. It is best to keep them as private instance functions at this moment.

File: packages/quantic/force-app/examples/main/lwc/exampleQuanticRefineModalContent/exampleQuanticRefineModalContent.js-meta.xml
Comment (erocheleau) [inline packages/quantic/force-app/examples/main/lwc/exampleQuanticRefineModalContent/exampleQuanticRefineModalContent.js-meta.xml:11]: Do we really want to expose in all of these targets? keep in mind this adds this example component in all the page builder for these environment

File: packages/quantic/force-app/main/default/lwc/quanticCategoryFacet/quanticCategoryFacet.js
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticCategoryFacet/quanticCategoryFacet.js:200]:
```suggestion
      facetId: this.facet?.state?.facetId,
      element: this.template.host
```

maybe?

Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticCategoryFacet/quanticCategoryFacet.js:200]: oh... why do we need a reference towards the template.host now?

File: packages/quantic/force-app/main/default/lwc/quanticHeadlessLoader/quanticHeadlessLoader.js
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticHeadlessLoader/quanticHeadlessLoader.js:270]: This seems more or less like a weird workaround no?

File: packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.css
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.css:5]: Just FYI, I don't know what we've decided to do here, but there is already slds classes to do that afaik

File: packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.html
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.html:70]: Just for readability, can we put timeFrameFacet higher than the default one?

File: packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.js
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.js:43]: Is there a way other facets could export those? Would we want to handle it like that instead?

**Review by erocheleau — APPROVED** [review]
There are a couple of fundamental questions raised I think and I'll let @nathanlb team decide if they are okay with the global changes, example of passing the this.template.host as a prop, but otherwise all the code looks great! 👍

File: packages/quantic/force-app/examples/main/lwc/exampleQuanticRefineModalContent/exampleQuanticRefineModalContent.js-meta.xml
Comment (mmitiche) [reply packages/quantic/force-app/examples/main/lwc/exampleQuanticRefineModalContent/exampleQuanticRefineModalContent.js-meta.xml:11]: probably only the community page is needed here true.

File: packages/quantic/force-app/main/default/lwc/quanticCategoryFacet/quanticCategoryFacet.js
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticCategoryFacet/quanticCategoryFacet.js:200]: I'm passing the template.host here to save all the attribute of the facet in the Quantic store which give us the ability to duplicate the facets in the modal with the same values for all attributes to get the exact same facets.

File: packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.html
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.html:70]: Sure!

File: packages/quantic/force-app/main/default/lwc/quanticHeadlessLoader/quanticHeadlessLoader.js
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticHeadlessLoader/quanticHeadlessLoader.js:270]: This was the only workaround I found to be able to create the exact duplicates of each facet, any ideas for a better alternative solution?

**Comment by lvu285** [comment]
Are you sure we still want to display "SORT component" as a dropdown inside the model?

**Comment by nathanlb** [comment]
> Are you sure we still want to display "SORT component" as a dropdown without formatting it to have similar style with other Filters inside the model? image

the mock-ups show a good solution for this. adding css to lwc can be tough due to the shadow dom but hopefully it won't be too bad.

**Comment by mmitiche** [comment]
> Are you sure we still want to display "SORT component" as a dropdown without formatting it to have similar style with other Filters inside the model?
> image
Good point I think it will be great to format the sort component to have a similar style to fit with the other filters👌
I'm just wondering if we need to do that now or we need to do it in another iteration since our main focus now is in the Hosted Insight Panel  which will have a QuanticRefineModalContent without the sort component.
I would like to hear the team's opinion on this.

**Comment by nathanlb** [comment]
> Good point I think it will be great to format the sort component to have a similar style to fit with the other filters👌 I'm just wondering if we need to do that now or we need to do it in another iteration since our main focus now is in the Hosted Insight Panel which will have a QuanticRefineModalContent without the sort component. I would like to hear the team's opinion on this.

It can be done separately just create a user story jira and you can add it to this epic https://coveord.atlassian.net/browse/SVCC-76

**Comment by mmitiche** [comment]
> It can be done separately just create a user story jira and you can add it to this epic https://coveord.atlassian.net/browse/SVCC-76

Perfect! Thank you!

File: packages/quantic/force-app/examples/main/lwc/exampleQuanticRefineModalContent/exampleQuanticRefineModalContent.html
Comment (lvu285) [inline packages/quantic/force-app/examples/main/lwc/exampleQuanticRefineModalContent/exampleQuanticRefineModalContent.html:78]: for testing, I suggest to setup the page where `The Quantic-refine-modal is passed in the content slot of the Quantic-modal Component` _ like the way you show me, it's safer, as these 2 components should be used together (?)

**Comment by lvu285** [comment]
general question:
do we plan to make all new Quantic components 'Accessibility' compliant?

File: packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.js
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.js:43]: We can the properties as static properties in each facets to get something like this:
static
and then then we can use them in the QuanticRefineModalContent component as follows:
`const NUMERIC_FACET_PROPERTIES = QuanticNumericFacet.attributes;`

What do you think?

File: packages/quantic/force-app/examples/main/lwc/exampleQuanticRefineModalContent/exampleQuanticRefineModalContent.html
Comment (mmitiche) [reply packages/quantic/force-app/examples/main/lwc/exampleQuanticRefineModalContent/exampleQuanticRefineModalContent.html:78]: Actually the two components will be tested together in the tests of the [QuanticRefineToggle component](https://github.com/coveo/ui-kit/blob/055987fec3f4354a4143f7e8410514da89a65c40/packages/quantic/force-app/main/default/lwc/quanticRefineToggle/quanticRefineToggle.html#L17-L40). this component is the button that will be used to open the refine modal.
I will open[ the PR of the QuanticRefineToggle](https://github.com/coveo/ui-kit/pull/1970) after merging this one.

File: packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.js
Comment (nathanlb) [inline packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.js:46]: instead of "a modal" i think it would be better to directly mention the QuanticModal component you built for this purpose :)

File: packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.js
Comment (nathanlb) [inline packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.js:98]:
```suggestion
   * Gather all facets registered in the Quantic store.
```

or 

```suggestion
   * Gather all registered facets from the Quantic store.
```

File: packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.html
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.html:23]: No, we're not removing the other facets. For example, in ExampleSearch, we'll just hide the original facets with a bit of CSS when the search interface is open in a narrow screen, and instead show the facets of refined modal.
but both sets of facets are always there at all time and both communicate with the engine, not sure how we can improve that :/

Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticRefineModalContent/quanticRefineModalContent.html:23]: Yes, from an accessibility standpoint the originals are hidden and the screen reader can't see them.
Anything having the the  CSS property : `display: none` will be hidden from an Accessibility standpoint.

**Review by nathanlb — APPROVED** [review]
GG!

**Pull request merged**
